baremetal/cluster_manager_test.go

```diff
@@ -0,0 +1,412 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package baremetal
+
+import (
+	"context"
+	"testing"
+
+	_ "github.com/go-logr/logr"
+	bmh "github.com/metal3-io/baremetal-operator/pkg/apis/metal3/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	_ "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/klogr"
+	infrav1 "sigs.k8s.io/cluster-api-provider-baremetal/api/v1alpha2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var bmcSpec = &infrav1.BareMetalClusterSpec{
+	APIEndpoint: "http://192.168.111.249:6443",
+}
+
+var bmcOwnerRef = &metav1.OwnerReference{
+	APIVersion: clusterv1.GroupVersion.String(),
+	Kind:       "Cluster",
+	Name:       clusterName,
+}
+
+const (
+	clusterName          = "testCluster"
+	baremetalClusterName = "testBaremetalCluster"
+	namespaceName        = "testNameSpace"
+)
+
+func TestNewClusterManagerCreate(t *testing.T) {
+
+	testCases := map[string]struct {
+		c             client.Client
+		BMCluster     *infrav1.BareMetalCluster
+		Cluster       *clusterv1.Cluster
+		ExpectSuccess bool
+	}{
+		"Cluster and BMCluster exist": {
+			Cluster:       newCluster(clusterName),
+			BMCluster:     newBareMetalCluster(baremetalClusterName, bmcOwnerRef, bmcSpec, nil),
+			ExpectSuccess: true,
+		},
+		"Cluster exists, BMCluster empty": {
+			Cluster:       newCluster(clusterName),
+			BMCluster:     &infrav1.BareMetalCluster{},
+			ExpectSuccess: false,
+		},
+		"Cluster empty, BMCluster exists": {
+			Cluster:       &clusterv1.Cluster{},
+			BMCluster:     newBareMetalCluster(baremetalClusterName, bmcOwnerRef, bmcSpec, nil),
+			ExpectSuccess: true,
+		},
+		"Cluster and BMCluster are nil": {
+			Cluster:       nil,
+			BMCluster:     nil,
+			ExpectSuccess: false,
+		},
+		"Cluster exists and BMCluster is nil": {
+			Cluster:       newCluster(clusterName),
+			BMCluster:     nil,
+			ExpectSuccess: false,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Logf("## TC-%s ##", name)
+
+		objects := []runtime.Object{
+			tc.Cluster, tc.BMCluster,
+		}
+
+		if tc.Cluster == nil || tc.BMCluster == nil {
+			tc.c = fakeclient.NewFakeClientWithScheme(setupScheme())
+		} else {
+			tc.c = fakeclient.NewFakeClientWithScheme(setupScheme(), objects...)
+		}
+
+		clusterMgr, err := newClusterManager(tc.c, tc.Cluster, tc.BMCluster, klogr.New())
+
+		if err != nil {
+			if tc.ExpectSuccess {
+				t.Error(err)
+			}
+		}
+
+		if clusterMgr == nil {
+			if tc.ExpectSuccess {
+				t.Error(err)
+			}
+		} else {
```

Comment:
This is repeated four times. you can make a function out of it and call it in each test function

Reply:
done